ivy/functional/ivy/experimental/manipulation.py

```diff
@@ -1811,38 +1845,45 @@ def put_along_axis(
     >>> arr = ivy.array([[4, 3, 5], [1, 2, 1]])
     >>> indices = ivy.array([[0, 1, 1], [2, 0, 0]])
     >>> values = ivy.array([[9, 8, 7], [6, 5, 4]])
-    >>> ivy.put_along_axis(arr, indices, values, 1, mode='clip')
+    >>> ivy.put_along_axis(arr, indices, values, 1, mode='assign')
+    >>> print(arr)
+    ivy.array([[9, 7, 5],
+               [4, 2, 6]])
+
+    >>> arr = ivy.array([[10, 30, 20], [60, 40, 50]])
+    >>> axis = 1
+    >>> indices = ivy.argmax(arr, axis=axis, keepdims=True)
+    >>> value = 100
+    >>> ivy.put_along_axis(arr, indices, value, axis, mode='add')
     >>> print(arr)
-    ivy.array([[3, 7, 5],
-               [6, 4, 1]])
+    ivy.array([[ 10, 130, 20],
+               [ 160, 40, 50]])
     """
-    if out is None:
-        out = ivy.zeros_like(arr)
+    arr_shape = arr.shape
 
-    indices = ivy.expand_dims(indices, axis=axis)
-    values = ivy.expand_dims(values, axis=axis)
+    # array containing all flat indices
+    arr_ = ivy.arange(0, _numel(arr_shape)).reshape(arr_shape)
 
-    stacked = ivy.concat((arr, values), axis=axis)
+    # use take_along_axis to get the queried indices
+    arr_idxs = ivy.take_along_axis(arr_, indices, axis)
 
-    sorted_indices = ivy.argsort(indices, axis=axis)
-    sorted_stacked = ivy.take_along_axis(stacked, sorted_indices, axis=axis)
+    # convert the flat indices to multi-D indices
+    arr_idxs = ivy.unravel_index(arr_idxs, arr_shape)
 
-    arr = ivy.where(
-        ivy.expand_dims(sorted_indices < arr.shape[axis], axis=axis),
-        sorted_stacked,
-        arr,
-    )
+    # stack the multi-D indices to bring them to scatter_nd format
+    arr_idxs = ivy.stack(arr_idxs, axis=-1).astype(ivy.int64)
 
-    if mode == "clip":
-        indices = ivy.clip(indices, 0, arr.shape[axis] - 1)
-    elif mode == "wrap":
-        indices = ivy.mod(indices, arr.shape[axis])
+    ret = ivy.scatter_nd(arr_idxs, values, reduction=mode, out=ivy.copy_array(arr))
```

Comment:
I probably shouldn't have used `copy_array` here now that I think about it.

Reply:
The way the `out` argument works in ivy functions is: If `out` is None the return is a new array, else the array provided by `out` is in-place updated with the return.
The backends and the compositional implementation need to be updated to do exactly that.